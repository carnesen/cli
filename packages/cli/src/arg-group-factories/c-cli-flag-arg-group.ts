import {
	CCliAbstractArgGroup,
	CCliArgGroupOptions,
} from '../c-cli-abstract-arg-group';
import { CCliUsageError } from '../c-cli-usage-error';

/** Options for {@link CCliFlagArgGroup} */
export type CCliFlagArgGroupOptions = Pick<
	CCliArgGroupOptions<false>,
	'description' | 'hidden'
>;

/** A `boolean`-valued command-line argument group
 * @example
 * ```plaintext
 * $ cli           // named flag "foo" parses `false`
 * $ cli --foo     // named flag "foo" parses `true`
 * $ cli --foo bar // usage error
 * ``` */
export class CCliFlagArgGroup extends CCliAbstractArgGroup<boolean, false> {
	public parse(args: string[] | undefined): boolean {
		// `args` `undefined` means that the named argument separator e.g. `--foo`
		// was not passed on the command line
		if (!args) {
			return false;
		}
		if (args.length > 0) {
			throw new CCliUsageError(`Unexpected argument "${args[0]}"`);
		}
		return true;
	}

	public static create(
		options: CCliFlagArgGroupOptions = {},
	): CCliFlagArgGroup {
		return new CCliFlagArgGroup({
			placeholder: '',
			required: false,
			hidden: options.hidden,
			description: options.description,
		});
	}
}
