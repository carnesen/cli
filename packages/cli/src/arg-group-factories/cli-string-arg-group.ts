import { CliArgGroup } from '../cli-arg-group';
import { CliUsageError } from '../cli-usage-error';

/** Options for [[`CliStringArgGroup`]] */
export type CliStringArgGroupOptions = {
	/** See [[`CliArgGroup.required`]] */
	required?: boolean;

	/** See [[`CliArgGroup.description`]] */
	description?: string;

	/** See [[`CliArgGroup.hidden`]] */
	hidden?: boolean;

	/** See [[`CliArgGroup.placeholder`]] defaulting to "\<str\>" */
	placeholder?: string;
};

/** A factory for required `string`-valued [[`CliArgGroup`]]s */
function CliStringArgGroup(
	options: CliStringArgGroupOptions & { required: true },
): CliArgGroup<string, true>;

/** A factory for optional `string | undefined`-valued [[`CliArgGroup`]]s */
function CliStringArgGroup(
	options?: CliStringArgGroupOptions,
): CliArgGroup<string | undefined, false>;

// Implementation
function CliStringArgGroup(
	options: CliStringArgGroupOptions = {},
): CliArgGroup<string | undefined> {
	const {
		required = false,
		description,
		placeholder = '<str>',
		hidden = false,
	} = options;
	const argGroup: CliArgGroup<string | undefined> = {
		hidden,
		placeholder,
		required,
		parse(args) {
			if (!args) {
				return undefined;
			}

			if (args.length > 1) {
				throw new CliUsageError(`Expected just one ${placeholder}`);
			}

			if (args.length === 0) {
				throw new CliUsageError(`Expected a ${placeholder}`);
			}

			return args[0];
		},
		description,
	};
	return argGroup;
}

export { CliStringArgGroup };
