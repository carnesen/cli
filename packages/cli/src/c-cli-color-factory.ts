import { cCliProcessFactory } from './c-cli-process';
import { CCliColor } from './c-cli-color';
import { CCliAnsiColor } from './c-cli-ansi-color';
import { CCliNoopColor } from './c-cli-noop-color';

/** A factory for {@link CCliColor}s
 * @param ansi If `true`, return a `CCliAnsiColor` instance. If `false`, return a
 * `CCliNoopColor` instance. If `undefined`, return a `CliAnsiColor` instance if
 * stdout and stderr are both a TTY or a `CliNoopColor` instance otherwise. */
export function cCliColorFactory(ansi?: boolean): CCliColor {
	switch (ansi) {
		case true: {
			return CCliAnsiColor.create();
		}
		case false: {
			return CCliNoopColor.create();
		}
		case undefined:
		default: {
			const { stderr, stdout } = cCliProcessFactory();
			return stdout.isTTY && stderr.isTTY
				? CCliAnsiColor.create()
				: CCliNoopColor.create();
		}
	}
}
