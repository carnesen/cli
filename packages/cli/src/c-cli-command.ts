import {
	InferParsedValueFromCCliArgGroup,
	CCliArgGroup,
} from './c-cli-arg-group';
import { CCliLogger } from './c-cli-logger';
import { CCliColor } from './c-cli-color';
import { CCliDescription } from './c-cli-description';

/** Type of the `input` argument injected into injected into
 * {@link CCliCommandOptions.action} */
export type CCliCommandActionInput<
	PositionalArgGroup extends CCliArgGroup,
	NamedArgGroups extends {
		[name: string]: CCliArgGroup;
	},
	DoubleDashArgGroup extends CCliArgGroup,
> = {
	color: CCliColor;
	doubleDashValue: InferParsedValueFromCCliArgGroup<DoubleDashArgGroup>;
	logger: CCliLogger;
	namedValues: {
		[K in keyof NamedArgGroups]: InferParsedValueFromCCliArgGroup<
			NamedArgGroups[K]
		>;
	};
	positionalValue: InferParsedValueFromCCliArgGroup<PositionalArgGroup>;
};

/** Options for creating a **@carnesen/cli** command */
export type CCliCommandOptions<
	PositionalArgGroup extends CCliArgGroup = CCliArgGroup,
	NamedArgGroups extends {
		[name: string]: CCliArgGroup;
	} = {
		[name: string]: CCliArgGroup;
	},
	DoubleDashArgGroup extends CCliArgGroup = CCliArgGroup,
> = {
	// Mandatory options

	/** Function or async function that implements the command */
	action: (
		input: CCliCommandActionInput<
			PositionalArgGroup,
			NamedArgGroups,
			DoubleDashArgGroup
		>,
	) => any;

	/** Identifier for this command in command-line usage */
	name: string;

	// Optional options

	/** A sentence or two about this command for command-line usage or a function
	 * returning the description text */
	description?: CCliDescription;

	/** A {@link CCliArgGroup} for the arguments after a lone "--" */
	doubleDashArgGroup?: DoubleDashArgGroup;

	/** If `true`, don't show this command in command-line usage */
	hidden?: boolean;

	/** A {@link CCliArgGroup} for the arguments passed as "--name value" */
	namedArgGroups?: NamedArgGroups;

	/** A {@link CCliArgGroup} for the arguments before the first separator argument */
	positionalArgGroup?: PositionalArgGroup;
};

/** A **@carnesen/cli** command-line interface (CLI) command */
export class CCliCommand<
	PositionalArgGroup_ extends CCliArgGroup = CCliArgGroup,
	NamedArgGroups_ extends {
		[name: string]: CCliArgGroup;
	} = {
		[name: string]: CCliArgGroup;
	},
	DoubleDashArgGroup_ extends CCliArgGroup = CCliArgGroup,
> {
	protected constructor(
		protected readonly options: CCliCommandOptions<
			PositionalArgGroup_,
			NamedArgGroups_,
			DoubleDashArgGroup_
		>,
	) {}

	public action(
		input: CCliCommandActionInput<
			PositionalArgGroup_,
			NamedArgGroups_,
			DoubleDashArgGroup_
		>,
	): any {
		return this.options.action(input);
	}

	public get description(): CCliDescription {
		return this.options.description;
	}

	public get doubleDashArgGroup(): DoubleDashArgGroup_ | undefined {
		return this.options.doubleDashArgGroup;
	}

	public get hidden(): boolean {
		return this.options.hidden ?? false;
	}

	public get name(): string {
		return this.options.name;
	}

	public get namedArgGroups(): NamedArgGroups_ | undefined {
		return this.options.namedArgGroups;
	}

	public get positionalArgGroup(): PositionalArgGroup_ | undefined {
		return this.options.positionalArgGroup;
	}

	/** Factory for **@carnesen/cli** command-line interface (CLI) commands */
	public static create<
		PositionalArgGroup extends CCliArgGroup = CCliArgGroup<unknown, false>,
		NamedArgGroups extends {
			[name: string]: CCliArgGroup;
		} = {
			[name: string]: CCliArgGroup;
		},
		DoubleDashArgGroup extends CCliArgGroup = CCliArgGroup<unknown, false>,
	>(
		options: CCliCommandOptions<
			PositionalArgGroup,
			NamedArgGroups,
			DoubleDashArgGroup
		>,
	): CCliCommand<PositionalArgGroup, NamedArgGroups, DoubleDashArgGroup> {
		return new CCliCommand<
			PositionalArgGroup,
			NamedArgGroups,
			DoubleDashArgGroup
		>(options);
	}
}
