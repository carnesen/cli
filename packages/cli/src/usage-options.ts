import { CliColor } from './cli-color';

export interface UsageOptions {
	color: CliColor;
	columns: number;
	indentation: string;
}
