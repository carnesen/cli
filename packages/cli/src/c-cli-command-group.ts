import { CCliCommand } from './c-cli-command';
import { CCliAnyDescription } from './c-cli-description';

/** Options for {@link CliCommandGroup} */
export type CCliCommandGroupOptions = {
	/** Name of this command group */
	name: string;

	/** A short description for command-line usage */
	description?: CCliAnyDescription;

	/** If `true`, hide these commands in command-line usage */
	hidden?: boolean;

	/** The {@link ICliCommandGroup`]]s and/or [[`ICliCommand}s in this group */
	subcommands: (CCliCommandGroup | CCliCommand<any, any, any>)[];
};

export class CCliCommandGroup {
	protected constructor(public readonly options: CCliCommandGroupOptions) {}

	public static create(options: CCliCommandGroupOptions): CCliCommandGroup {
		return new CCliCommandGroup(options);
	}
}
