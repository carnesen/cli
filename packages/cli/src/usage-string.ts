import { CliTree } from './cli-tree';
import { CLI_COMMAND_GROUP } from './cli-command-group';
import { usageForCommandGroup } from './usage-for-command-group';
import { CLI_COMMAND } from './cli-command';
import { usageForCommand } from './usage-for-command';
import { UsageOptions } from './usage-options';

export function usageFactory(tree: CliTree, options: UsageOptions): string {
	const { current, parents } = tree;

	let lines: string[] = [];
	if (current.kind === CLI_COMMAND_GROUP) {
		lines = usageForCommandGroup({ current, parents }, options);
	} else if (current.kind === CLI_COMMAND) {
		lines = usageForCommand({ current, parents }, options);
	} else {
		throw new Error('Unexpected kind');
	}
	return lines.join('\n');
}
