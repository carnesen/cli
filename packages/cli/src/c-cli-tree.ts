import { CCliCommandGroup } from './c-cli-command-group';
import { CCliCommand } from './c-cli-command';
import { CCliAbstractArgGroup } from './c-cli-abstract-arg-group';

/** Root of a command tree */
export type CCliRoot =
	| CCliCommandGroup
	| CCliCommand<CCliAbstractArgGroup, any, CCliAbstractArgGroup>;

/** Data structure representing a node in a command tree */
export type CCliTree = {
	current: CCliRoot;
	parents: CCliCommandGroup[];
};

/** Data structure representing a leaf node in a command tree */
export type CCliLeaf = {
	current: CCliCommand;
	parents: CCliCommandGroup[];
};
