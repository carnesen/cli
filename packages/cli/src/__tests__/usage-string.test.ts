import { runAndCatchSync } from '@carnesen/run-and-catch';
import { usageFactory } from '../usage-string';
import { CliStringArgGroup } from '../arg-group-factories/cli-string-arg-group';
import { CCliCommand } from '../c-cli-command';
import { CCliTree } from '../c-cli-tree';
import { UsageOptions } from '../usage-options';
import { cCliColorFactory } from '../c-cli-color-factory';
import { CCliCommandGroup } from '../c-cli-command-group';

const DESCRIPTION = 'A string message please';

const messageArgGroup = CliStringArgGroup({
	description: DESCRIPTION,
});
const positionalArgGroup = CliStringArgGroup({
	description: 'A word',
	placeholder: '<word>',
});
const doubleDashArgGroup = CliStringArgGroup({
	description: 'Another word',
	required: true,
});

const current = CCliCommand.create({
	name: 'echo',
	positionalArgGroup,
	namedArgGroups: {
		message: messageArgGroup,
	},
	doubleDashArgGroup,
	action(foo) {
		return foo;
	},
});

const commandGroup = CCliCommandGroup.create({
	name: 'cli',
	description: 'This is a CLI',
	subcommands: [current],
});

const options: UsageOptions = {
	color: cCliColorFactory(),
	columns: 100,
	indentation: '',
};

describe(usageFactory.name, () => {
	it('Creates a usage string for a command group', () => {
		const usageString = usageFactory(
			{ current: commandGroup, parents: [] },
			options,
		);
		expect(usageString).toMatchSnapshot();
	});

	it('Creates a usage string for a command without a parent', () => {
		const usageString = usageFactory(
			{
				current: current as CCliTree['current'],
				parents: [],
			},
			options,
		);
		expect(usageString).toMatch(DESCRIPTION);
		expect(usageString).toMatchSnapshot();
	});

	it('Creates a usage string for a command without a parent command group', () => {
		const usageString = usageFactory(
			{
				current: current as CCliTree['current'],
				parents: [commandGroup],
			},
			options,
		);
		expect(usageString).toMatchSnapshot();
	});

	it('Does not write usage for named argGroups if there are none', () => {
		const fooCommand = CCliCommand.create({ name: 'foo', action() {} });
		const usageString = usageFactory(
			{ current: fooCommand, parents: [] },
			options,
		);
		expect(usageString).toMatchSnapshot();
	});

	it('Throws "unexpected kind" if passed an object of unknown kind', () => {
		const exception = runAndCatchSync(
			usageFactory,
			{ current: {} } as any,
			{} as any,
		);
		expect(exception.message).toMatch(/unexpected kind/i);
	});
});
