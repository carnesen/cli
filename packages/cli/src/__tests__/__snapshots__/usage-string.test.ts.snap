// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usageFactory Creates a usage string for a command group 1`] = `
"Usage: cli <subcommand> ...

This is a CLI

Subcommands:

echo"
`;

exports[`usageFactory Creates a usage string for a command without a parent 1`] = `
"Usage: echo [<word>] [<named args>] -- <str>

Positional arguments:

<word> : A word

Named arguments:

[--message <str>] : A string message please

\\"Double dash\\" arguments:

<str> : Another word"
`;

exports[`usageFactory Creates a usage string for a command without a parent command group 1`] = `
"Usage: cli echo [<word>] [<named args>] -- <str>

Positional arguments:

<word> : A word

Named arguments:

[--message <str>] : A string message please

\\"Double dash\\" arguments:

<str> : Another word"
`;

exports[`usageFactory Does not write usage for named argGroups if there are none 1`] = `
"Usage: foo
"
`;
